### DISCUSSION TITLE
reading files trhough NFS(version 3,4.1) get stuck

### DISCUSSION THREAD

**Discussion by avishayhatton** [openzfs/zfs#12370] [General]
Hi all,
a strange thing with new version (2.0.4/2.05):
Linux platform(aarch64)
1. we were working with 0.8.5 version as a shared storage - NFS 4.1 - read/write for 3 months.
2.  we upgraded to 2.0.4(remove ZFS completley, reboot, then build the new zfs) 
3. imported the pool back and restart nfs-server + zpool upgrade -a
4. when reading a file locally from a dataset on the ZFS storage, everything is ok. you can read the contents 
5. when you read the file using NFS protocol(tried Version3 and version 4.1) - the read operation get stuck(less/cat tools)
cat /mnt/folder/ap_dro_7p5t_all_VT.v

only CTRL-C helps to get back to prompt
6. the same file can be read from local ZFS storage.
7. some files in the same folder can be read from remote(NFS) and some not(get stuck with cat/less)
8. when i downgrade back to 0.8.5 (imported with read only mode) - all files can be read from remote(nfs)

Thanks
Avishay

**Comment by rincebrain** [comment]
What OS/release/kernel version the server and client are running may be relevant here,

Does anything show up in the server or client's dmesg?

You say 2.0.4/2.0.5 then only mention 2.0.4 in your story - do you mean that you have also since tested 2.0.5?

**Reply by avishayhatton** [reply]
storage side 
lsb_release -a
LSB Version:    :core-4.1-aarch64:core-4.1-noarch
Distributor ID: Amazon
Description:    Amazon Linux release 2 (Karoo)
Release:        2
Codename:       Karoo


uname -r
4.14.232-176.381.amzn2.aarch64

build with DKMS

client:  - X86_64 - centos 7.9(3.10.0-1062.1.1.el7.x86_64) or Amazon Linux 2(4.14.232-176.381.amzn2.x86_64)


no recent messages using "dmesg -Tx" and checking /var/log/messages.

checked 2.0.4 and 2.0.5
haven't tested 2.1 yet

filename:       /lib/modules/4.14.232-176.381.amzn2.aarch64/extra/zfs/zfs/zfs.ko
version:        2.0.5-1
license:        CDDL
author:         OpenZFS
description:    ZFS
alias:          devname:zfs


p.s.
tested and get the same behavior on 3 different sotrage instances.

**Reply by avishayhatton** [reply]
Hi, 
another update,
1. we get the same issue, with zfs-2.1
2. copying over the file with a new name and change ownership/permissions - didn't change a thing.

**Reply by rincebrain** [reply]
1 doesn't really surprise me
I don't understand what you're trying to communicate with 2 - copied from where, to where? Are you saying you can reliably reproduce this with a certain file even with a new copy of the file? On the same or different datasets?

**Reply by avishayhatton** [reply]
I tried to copy the file, i could not  read through nfs with a new name and chnaged it's ownership and permissions, to see if it changes anything(able to read it from NFS).  it didn't help.(cat/less file_name gets stuck)

of course, reading the file from the local ZFS storage is fine.

**Reply by avishayhatton** [reply]
i wanted to see if i copy a file i could not read through NFS with a new name + changed ownership/permission, would change anything.
it didn't work.less/cat get stuck with  the new file name as well.
of course, reading the file from the local ZFS storage is fine.

**Reply by rincebrain** [reply]
Does anything show up in dmesg when this happens?

Can you tell me anything about the file? How large it is, whether the contents is lots of text or compressed video or /dev/urandom output, what compression algorithm is in use on the dataset, is encryption enabled and if so what algorithm...

If whatever the file is would allow it, since you said you could reproduce this on multiple machines, sharing it would probably make it easier to figure out what's going on. If you can't, that's perfectly fine, but it would likely speed things up.

**Reply by avishayhatton** [reply]
1. dmesg or tail -f /var/log/messages is clean(storage side)
2. the file is a text file(it is being compressed by ZFS lz4)
3. tried also zstd(created a new data set and copied the data over to the new data set)
3. Size: 41k(on disk) but original size is 345k(ls -lha file_name), record size is default=128k
4. no encryption enabled
5. i will try to re-produce it on intel X86_64

**Comment by avishayhatton** [comment]
Hi,
tested with 0.8.5.1 on x86_64

[root@duzfs-test2 ~]# arch
x86_64

filename:       /lib/modules/4.14.238-182.421.amzn2.x86_64/extra/zfs/zfs/zfs.ko
version:        0.8.5-1
1.  synced the data using zfs send/recieve (from the ZFS storage with version 2.1)
2. exported the data set using nfs
3. mounted from a client using nfs 3/4.1
4. all files can be read(simple loop to read all files in the same folder) 

next step: upgrade to 2.0.5

1. removed zfs rpms and rebooted
3. downloaded tar(2.0.5) and  compiled (rpms, rpms-dkms)
4. rebooted
3. import without upgrading pool version
4. reading get stuck.(cat, less)
5. dmesg -Tx, tail -f /var/log/messages - clean

**Reply by rincebrain** [reply]
Neat.

Can you tell me the properties on the dataset (`zfs get all [dataset]`)? Are there any xattrs set? Can you share the output of `zdb -Ovv [dataset] [relative path to file from the root of the dataset]`? Does this also happen if you mount the NFS export locally on the server (e.g. `mount -t nfs localhost:/path /local/path`) and try reading it?

**Reply by avishayhatton** [reply]
1. mounting localy from ZFS storage have the same stuck issue.
2.  tested with both Raidz1 and raid0(1 disk)



zpool create -f v01 -o ashift=12 -O redundant_metadata=most -O atime=off -O xattr=sa -O sync=disabled -O compression=lz4
by default we set xattr=sa


v01/test  type                  filesystem             -
v01/test  creation              Wed Jul 21 12:15 2021  -
v01/test  used                  12.6G                  -
v01/test  available             83.8G                  -
v01/test  referenced            12.6G                  -
v01/test  compressratio         3.55x                  -
v01/test  mounted               yes                    -
v01/test  quota                 none                   default
v01/test  reservation           none                   default
v01/test  recordsize            128K                   default
v01/test  mountpoint            /v01/test             default
v01/test  sharenfs              off                    default
v01/test  checksum              on                     default
v01/test  compression           lz4                    inherited from v01
v01/test  atime                 off                    inherited from v01
v01/test  devices               on                     default
v01/test  exec                  on                     default
v01/test  setuid                on                     default
v01/test  readonly              off                    default
v01/test  zoned                 off                    default
v01/test  snapdir               hidden                 default
v01/test  aclmode               discard                default
v01/test  aclinherit            restricted             default
v01/test  createtxg             433                    -
v01/test  canmount              on                     default
v01/test  xattr                 sa                     inherited from v01
v01/test  copies                1                      default
v01/test  version               5                      -
v01/test  utf8only              off                    -
v01/test  normalization         none                   -
v01/test  casesensitivity       sensitive              -
v01/test  vscan                 off                    default
v01/test  nbmand                off                    default
v01/test  sharesmb              off                    default
v01/test  refquota              none                   default
v01/test  refreservation        none                   default
v01/test  guid                  12327002872035453633   -
v01/test  primarycache          all                    default
v01/test  secondarycache        all                    default
v01/test  usedbysnapshots       0B                     -
v01/test  usedbydataset         12.6G                  -
v01/test  usedbychildren        0B                     -
v01/test  usedbyrefreservation  0B                     -
v01/test  logbias               latency                default
v01/test  objsetid              134                    -
v01/test  dedup                 off                    default
v01/test  mlslabel              none                   default
v01/test  sync                  disabled               inherited from v01
v01/test  dnodesize             legacy                 default
v01/test  refcompressratio      3.55x                  -
v01/test  written               0                      -
v01/test  logicalused           44.4G                  -
v01/test  logicalreferenced     44.4G                  -
v01/test  volmode               default                default
v01/test  filesystem_limit      none                   default
v01/test  snapshot_limit        none                   default
v01/test  filesystem_count      none                   default
v01/test  snapshot_count        none                   default
v01/test  snapdev               hidden                 default
v01/test  acltype               off                    default
v01/test  context               none                   default
v01/test  fscontext             none                   default
v01/test  defcontext            none                   default
v01/test  rootcontext           none                   default
v01/test  relatime              off                    default
v01/test  redundant_metadata    most                   inherited from v01
v01/test  overlay               on                     default
v01/test  encryption            off                    default
v01/test  keylocation           none                   default
v01/test  keyformat             none                   default
v01/test  pbkdf2iters           0                      default
v01/test  special_small_blocks  0                      default

**Reply by avishayhatton** [reply]
Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
      2009    1   128K    512      0     512    512    0.00  ZFS plain file
                                               193   bonus  System attributes
        dnode flags: USED_BYTES USERUSED_ACCOUNTED USEROBJUSED_ACCOUNTED
        dnode maxblkid: 0
        target  ap_dro_7p5t_6.00a
        uid     2032
        gid     2004
        atime   Wed Jul 14 17:53:46 2021
        mtime   Sun May 19 11:14:28 2019
        ctime   Wed Jul 14 18:19:45 2021
        crtime  Wed Jul 14 17:53:46 2021
        gen     16947
        mode    120777
        size    17
        parent  477
        links   1
        pflags  40800000104
Indirect blocks:

**Reply by avishayhatton** [reply]
also created a new dataset with xattr=on and only then - copied the data over.
same Hung issue when reading some files.

**Reply by rincebrain** [reply]
And just to be clear, how large is the file you're listing now? Because it does not look like the 345k file you mentioned before.

**Reply by avishayhatton** [reply]
du -sh havishay-ap_dro_7p5t_all_VT.v
41K     havishay-ap_dro_7p5t_all_VT.v

 ls -hal havishay-ap_dro_7p5t_all_VT.v
-rwxr-xr-x 1 havishay anpa 345K Jul 20 11:10 havishay-ap_dro_7p5t_all_VT.v

**Reply by avishayhatton** [reply]
1. My mistake - i followed a soft link. Now i am using the real folder
2. also, i already checked using the full path(no soft link involved - and it gets stuck)
    Object  lvl   iblk   dblk  dsize  dnsize  lsize   %full  type
     75528    2   128K   128K    40K     512   384K  100.00  ZFS plain file
                                               176   bonus  System attributes
        dnode flags: USED_BYTES USERUSED_ACCOUNTED USEROBJUSED_ACCOUNTED
        dnode maxblkid: 2
        uid     2491
        gid     2001
        atime   Tue Jul 20 11:10:30 2021
        mtime   Tue Jul 20 11:10:30 2021
        ctime   Tue Jul 20 11:33:23 2021
        crtime  Tue Jul 20 11:10:30 2021
        gen     115004
        mode    100755
        size    352812
        parent  2023
        links   1
        pflags  840800000104
Indirect blocks:
               0 L1  0:6e63a8000:1000 20000L/1000P F=3 B=467/467 cksum=8cb0c8b930:1fbd80dab73cb:39787c30bbd9e34:598caf13dbc3899f
               0  L0 0:6e639f000:2000 20000L/2000P F=1 B=467/467 cksum=1bee61c8096:8253f61896133:16098fce0c7da8d0:faeaa299eb637540
           20000  L0 0:6e63a1000:4000 20000L/4000P F=1 B=467/467 cksum=31d06a20ebf:1a994ed6c71498:917773dd1267dfb4:f4ffa1d516e0e109
           40000  L0 0:6e63a5000:3000 20000L/3000P F=1 B=467/467 cksum=2804db7557a:d9f3de07f1d09:34b68ea93b5c4178:cfb7e2a0d16ddd1

                segment [0000000000000000, 0000000000060000) size  384K

**Reply by rincebrain** [reply]
That's curious.

It seems to reproduce readily if you set up an Amazon Linux 2 VM on VirtualBox and try reading a copy of the ZFS git tree (for me, it cropped up on config.log).

I wonder what's going on. Let's see...

**Reply by avishayhatton** [reply]
well,
tested it with EC2 instance + Centos 7.9. it is working fine with this release(centos 7.9 + zfs 2.0.5)

**Reply by rincebrain** [reply]
That makes sense, given what I'm about to remark.

It also happens with vanilla 4.14.240, even on loopback NFS mount. So that's entertaining.
